data/vector2.lean

```diff
@@ -96,12 +99,128 @@ def {u} mmap {m} [monad m] {α} {β : Type u} (f : α → m β) :
 
 end vector
 
-namespace equiv
+namespace vector
+
+universes u
+variables {n : ℕ}
+
+section traverse
+
+variables {f f' : Type u → Type u}
+variables [applicative f] [applicative f']
+
+open applicative functor
+open list (cons) nat
+
+def traverse_aux {α β : Type u} (g : α → f β) :
+  Π (x : list α), f (vector β x.length)
+| [] := pure vector.nil
+| (x::xs) := vector.cons <$> g x <*> traverse_aux xs
+
+protected def traverse {α β : Type u} (g : α → f β) :
+  vector α n → f (vector β n)
+ | ⟨v,Hv⟩ := cast (by rw Hv) $ traverse_aux g v
+
+protected def to_array {α : Type u} {n} : vector α n → array n α
+ | ⟨xs,h⟩ := cast (by rw h) xs.to_array
+
+variables [is_lawful_applicative f] [is_lawful_applicative f']
+variables {α β η : Type u}
+
+@[simp]
+protected lemma traverse_def (g : α → f β) (x : α) (xs : vector α n) :
+  vector.traverse g (x :: xs) = cons <$> g x <*> vector.traverse g xs :=
+begin
+  cases xs, simp!,
+  h_generalize _ : _ == i,
+  h_generalize _ : _ == j, subst n,
+  simp at *, subst i, subst j
+end
+
+protected lemma id_traverse (x : vector α n) :
+  vector.traverse id.mk x = x :=
+begin
+  cases x with x, subst n,
+  dsimp [vector.traverse,cast],
+  induction x with x xs, refl,
+  simp! [x_ih], refl
+end
+
+open function
+
+protected lemma comp_traverse (g : α → f β) (h : β → f' η) (x : vector α n) :
+  vector.traverse (comp.mk ∘ functor.map h ∘ g) x =
+  comp.mk (vector.traverse h <$> vector.traverse g x) :=
+begin
+  cases x with x,
+  dunfold vector.traverse, subst n, dsimp [cast],
+  induction x with x xs; simp! [cast,*] with functor_norm, refl,
+  congr' 2, ext, simp [function.comp,flip]
+end
+
+protected lemma map_traverse
+   (g : α → f' β) (f : β → η)
+   (x : vector α n) :
+  map f <$> vector.traverse g x = vector.traverse (functor.map f ∘ g) x :=
+begin
+  symmetry,
+  cases x with x, subst n, unfold vector.traverse cast,
+  induction x; simp! [*,cast,map,flip,function.comp,vector.map] with functor_norm, refl,
+  congr' 2, ext, cases x_1, refl
+end
+
+protected lemma traverse_map
```

Comment:
this and the next proof look very non-maintainable. You use a lot of `simp` and `simp!` without restricting it using `{ ... }`